sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java

```diff
@@ -33,25 +43,199 @@
+    }
+    kwargsMap.put(name, value);
+    return this;
+  }
+
+  /**
+   * Specifies keyword arguments for the Python cross-language transform. If invoked more than once,
+   * new keyword arguments map will be added to the previously prided keyword arguments.
+   *
+   * @return updated wrapper for the cross-language transform.
+   */
+  public ExternalPythonTransform<InputT, OutputT> withKwargs(Map<String, Object> kwargs) {
+    kwargsMap.putAll(kwargs);
+    return this;
+  }
+
+  /**
+   * Specifies keyword arguments as a Row objects.
+   *
+   * @param kwargs keyword arguments as a {@link Row} objects. An empty Row represents zero keyword
+   *     arguments.
+   * @return updated wrapper for the cross-language transform.
+   */
+  public ExternalPythonTransform<InputT, OutputT> withKwargs(Row kwargs) {
+    if (this.kwargsMap.size() > 0) {
+      throw new IllegalArgumentException("Kwargs were specified both directly and as a Row object");
+    }
+    this.providedKwargsRow = kwargs;
+    return this;
+  }
+
+  @VisibleForTesting
+  Row buildOrGetKvargsRow() {
+    if (providedKwargsRow != null) {
+      return providedKwargsRow;
+    } else if (kwargsMap.size() == 0) {
+      return Row.nullRow(Schema.of());
+    } else {
+      Schema schema =
+          generateSchemaFromFieldValues(
+              kwargsMap.values().toArray(), kwargsMap.keySet().toArray(new String[] {}));
+      return Row.withSchema(schema)
+          .addValues(convertComplexTypesToRows(kwargsMap.values().toArray()))
+          .build();
+    }
+  }
+
+  private static boolean isCustomType(java.lang.Class<?> type) {
+    boolean val =
+        !(ClassUtils.isPrimitiveOrWrapper(type)
+            || type == String.class
+            || Row.class.isAssignableFrom(type));
+    return val;
+  }
+
+  // If the custom type has a registered schema, we use that. OTherwise we try to register it using
+  // 'JavaFieldSchema'.
+  private Row convertCustomValue(Object value) {
+    SerializableFunction<Object, Row> toRowFunc;
+    try {
+      toRowFunc =
+          (SerializableFunction<Object, Row>) SCHEMA_REGISTRY.getToRowFunction(value.getClass());
+    } catch (NoSuchSchemaException e) {
+      SCHEMA_REGISTRY.registerSchemaProvider(value.getClass(), new JavaFieldSchema());
+      try {
+        toRowFunc =
+            (SerializableFunction<Object, Row>) SCHEMA_REGISTRY.getToRowFunction(value.getClass());
+      } catch (NoSuchSchemaException e1) {
+        throw new RuntimeException(e1);
+      }
+    }
+    return toRowFunc.apply(value);
+  }
+
+  private Object[] convertComplexTypesToRows(Object[] values) {
+    Object[] converted = new Object[values.length];
+    for (int i = 0; i < values.length; i++) {
+      Object value = values[i];
+      converted[i] = isCustomType(value.getClass()) ? convertCustomValue(value) : value;
+    }
+    return converted;
+  }
+
+  @VisibleForTesting
+  Row buildOrGetArgsRow() {
+    if (argsArray.length == 0) {
```

Comment:
Ideally generateSchemaFromFieldValues should handle the degenerate empty case just fine, so we don't need to add a special case. (Similarly above for the kwargs.)

Reply:
Updated and added tests.